README.md
# Hjalmion-The-Rise-Of-Fencor
My repository for my self made text-based adventure game, Hjalmion: The Rise Of Fencor, made in Python 3.X

This game has been coded by me, Knifesurge. If you have any requests on new things to put in or possibly enhancing existing code, please let me know! I am always available at hjalmionlabs@outlook.ca

# Updates
## (2019-03-22)
I have decided to do something new with this project since it has been sitting idle for the last ~~3~~ __5__ years.

I am going to be rewriting the vast majority of the codebase. This should help clean up the project and reduce the amount of lines the main file (Hjalmion.py) has (currently almost 3k lines!).

### Plans
I plan on making the game itself more flexible and customizable so that adding new/custom items/enemies/rooms is possible without rewriting large chunks of code. Also, I plan on moving away from using the *pickle* module, and moving towards saving using the JSON format. This seems (to me) the more reasonable choice, and follows with my plan on making the game more flexible (i.e, the save file is entirely editable).

The game will load Items, Enemies, and Rooms from respective JSON files. Again, this follows with the more flexible pursuit.




changelog.md
# 2019-03-22:
1. Updated `README` to match project pursuits.
2. Created `Hjalmion_new.py` to keep original and rewrite files separate. At the end of the project, `Hjalmion_new` will become just `Hjalmion`, and `Hjalmion` will become `Hjalmion_old`.
3. Added `.gitignore` file
4. Rewrote `Player.py` to only include one class definition
5. Added inventory to Player class itself (dict)
6. Added some helper methods to retrieve now 'private' attributes
7. Exp needed to advance to next level now doubles each level
8. Started fleshing out `Hjalmion_new.py`
9. Added `Rooms.py` to describe the rooms in the game. ~~Will be implemented as a Linked List~~ __Does not make sense to use a linked list structure__. Each room has N,NE,E,SE,S,SW,W,NW pointers to its adjacent rooms, effectively creating a map of the game world. One note however, is that it will not be possible to easily iterate through the list. One way to traverse is by historical addition to the list. That is, `Rooms` maintains a list (which acts like a set) of it's nodes in the order at which they are added, which allows it to be traversed in some way. 
10. Rooms now hold what enemies are inside of the room. Added helper methods to get the list of Enemies themselves, as well as pick a random Enemy from the list.
11. Rooms when printed now show what Rooms are adjacent to it, and which direction.
12. Added a `Utils` module that holds various functions to read and write data from JSON files.
13. Added *Created:* amd *Updated:* Comments at the head of each file in order to help track which files are new (part of the rewrite) and which are the original files (anything pre-2019)
14. Added a `data` folder that will hold all the json files for the game.
15. Created a `Enemies.json` file that holds all of the Enemies currently in the game. The layout in the file is the current standard for how to lay out the file and each Enemy's values.

# 2019-03-23:
1. Finished `Enemy#create_from_filedata`. Correctly creates enemies based on filedata.
2. Moved `Tests.py` to the main folder (instead of inside a tests folder) to test various aspects of the rewrite. Also added to .gitignore
3. Renamed `enemies.py` to `Enemy.py`. Used same format for `chance`, `items`, and `player`.

# 2019-03-24
1. Finished `Item#create_from_filedata`. Correctly creates items based on filedata.
2. Removed all `Item` definitions. Items created from filedata and used from lists inside the `Hjalmion` class.
3. Added __all__ Items to `Items.json`.
4. `Tests.py` now loads and prints enemies and items from filedata.

# 2019-03-27
1. `Utils#read_json` appends `None` to the end of the data cache after reading a file to separate the data. Added this in a previous update but forgot to mention it.
2. Added `Rooms.json` and the general format that the Rooms will follow.
3. Adjusted some of the `_Room` constructor code to follow the `Rooms.json` format.
4. Fixed `Rooms#insert` after changing the `_Room` constructor parameters.

# 2019-03-29
1. Started adding the `Room` definitions into `Rooms.json`.
2. Added `Ideas.txt` to track ideas and possible TODOs.
3. Added *dependencies* to the `Room` definitions. This will be a `list` of `Items` or `Conditions` in order to *access* the `Room` or to get a different *description*.

# 2019-04-10
1. Removed `victory` variable from `Player` since its not used.
2. Room movement will be handled by the `Player` class via a method named `move_room(dir:str)`. `Player#move_room` takes a char (str) as an argument, `dir`, representing the direction the Player should move. If a `Room` doesn't exist in that direction, `False` is returned and the `Player` is not moved.
3. `Rooms` is no longer a `Linked List` under the hood; it doesn't need it. Now just maintains a Python `List` of `Rooms`.
4. `Room` now properly prints its adjacent `Rooms` in its `__str__` method.
5. Added a `get_room_in_dir` method to `Room`, which gets an adjacent `Room` in a specified `dir` (str). If an invalid direction is provided, just returns itself.
6. ~~Cleaned up `Room#north`, `Room#south`, etc. Since in `_Room#__init__` it only adds rooms if they are not empty in `Rooms.json`, there is no need to check if it exists in `Room#__adj`, since it most certainly does.~~ Still needs it.
7. `Room#this` created as a helper function to `Room#get_room_in_dir` since it returns a function call.
8. Made a `docs` folder to hold related documentation for the `JSON` files for `Rooms`, `Enemies`, and `Items`.

Room.py
# Nick Mills
# Created: 2019-03-22
# Updated: 2019-04-10

from random import randint
from Utils import Utils

class _Room():
    def __init__(self, name="Unnamed Room",\
        description="No description given",dependencies=[],\
        enemies=[],adj={}):
        """
        """
        self.__name = name
        self.__description = description
        self.__dependencies = dependencies
        self.__enemies = enemies
        # Dict of adjacent rooms
        self.__adj = {}
        for k,v in adj.items():
            if v != "":
                self.__adj[k] = v

    def description(self):
        return self.__description

    def rooms(self) -> dict:
        """
        """
        return self.__adj

    def get_room_in_dir(self, dir : str):
        """
        """
        room = self.this
        valid_dirs = {'n':self.north,'s':self.south,\
            'e':self.east,'w':self.west,'ne':self.north_east,\
                'nw':self.north_west,'se':self.south_east,\
                'sw':self.south_west}
        if dir in valid_dirs.keys():
            room = valid_dirs[dir]
        return room()

    def enemies(self):
        return self.__enemies
    
    def this(self):
        return self

    def north(self):
        if 'n' in self.__adj.keys():
            n = self.__adj['n']
        else:
            n = None
        return n

    def north_east(self):
        if 'ne' in self.__adj.keys():
            ne = self.__adj['ne']
        else:
            ne = None
        return ne

    def north_west(self):
        if 'nw' in self.__adj.keys():
            nw = self.__adj['nw']
        else:
            nw = None
        return nw

    def south(self):
        if 's' in self.__adj.keys():
            s = self.__adj['s']
        else:
            s = None
        return s

    def south_east(self):
        if 'se' in self.__adj.keys():
            se = self.__adj['se']
        else:
            se = None
        return se

    def south_west(self):
        if 'sw' in self.__adj.keys():
            sw = self.__adj['sw']
        else:
            sw = None
        return sw

    def east(self):
        if 'e' in self.__adj.keys():
            e = self.__adj['e']
        else:
            e = None
        return e

    def west(self):
        if 'w' in self.__adj.keys():
            w = self.__adj['w']
        else:
            w = None
        return w

    def __repr__(self):
        _str = "{}\n{}\n{}\n".format(self.__name,'-'*(len(self.__name) + 2),\
            self.__description)

        count = 0
        _str += "\nAdjacent Rooms:\n"
        for k,v in self.__adj.items():
            if count % 2 != 0:
                _str += "{}: {}\n".format(k,v)
            else:
                _str += "{}: {}\t".format(k,v)
            count += 1
        print()
        return _str

    def __str__(self):
        return "{}\n{}\n{}\n".format(self.__name,'-'*(len(self.__name) + 2),\
            self.__description)

class Rooms():
    def __init__(self):
        self.__values = []
        self.__count = 0

    @staticmethod
    def create_from_filedata(rooms) -> None:
        """
        """
        data = Utils.get_data()
        index = 0
        while data[index] is not None:
            index += 1
        index += 1
        while data[index] is not None:
            index += 1
        room_data_start = index + 1
        room_data = data[room_data_start:]
        dicts = room_data[0]
        for _dict in dicts:
            rooms.insert(_dict)
        return

    def insert(self, data: dict):
        """
        """
        new_node = _Room(data['name'], data['description'], \
                data['dependencies'], data['enemies'], \
                data['adjacent'])
        self.__values.append(new_node)
        return    

    def first(self):
        """
        Returns the first Room added
        """
        return self.__values[0]

    def __str__(self):
        return f"Num Rooms: {len(self.__values)}"

    def __iter__(self):
        for room in self.__values:
            yield room